docs/Options.md
# Options usage

Command line arguments can be viewed by running:

```bash
docker run --rm dockupdater/dockupdater --help
```

All command line arguments can be substituted with an environment variable. All examples will be given as environment variables for a `docker run -e option=option_value`.

On a stack file (Docker swarm), you may set the environment variable like so:

```bash
version: "3.6"

services:
  dockupdater:
    image: dockupdater/dockupdater
    environment:
        - INTERVAL=30
        - LABEL=true
    deploy:
      labels:
        dockupdater.disable: "true"
      placement:
        constraints:
          - node.role == manager
```

> On stack mode, environment variables should not be enclosed in quotation marks.

***

* [Help](#help)
* [Version](#version)
* [Log Level](#log-level)
* [Scheduler](#scheduler)
  * [Cron](#cron)
  * [Interval](#interval)
  * [Run Once](#run-once)
* [Docker Specifics](#docker-specifics)
  * [Cleanup](#cleanup)
  * [Disable containers check](#disable-containers-check)
  * [Disable services check (swarm)](#disable-services-check-swarm)
  * [Docker Sockets](#docker-sockets)
  * [Docker TLS Verify](#docker-tls-verify)
  * [Label](#label)
  * [Latest](#latest)
  * [Recreate first](#recreate-first)
  * [Repository User](#repository-user)
  * [Repository Password](#repository-password)
  * [Start](#start)
  * [Stop](#stop)
  * [Wait time](#wait-time)
* [Notifications](#notifications)
  * [Notifiers](#notifiers)
  * [Skip start notification](#skip-start-notification)
  * [Template file](#template-file)

***

### Help

**Type:** `Boolean - Interrupting`  
**Command Line:**  `-h, --help`  

Shows the help message then exits.

### Version

**Type:** `Boolean - Interrupting`
**Command Line:**  `-v, --version`  

Shows the current version number then exits

### Log Level

**Type:** `String - Choice`  
**Command Line:**  `-l, --log-level`  
**Environment Variable:** `LOG_LEVEL`  
**Choices:**

* debug
* info
* warn
* error
* critical

**Default:** `info`  
**Example:** `-e LOG_LEVEL=info`  

Sets your logging verbosity level.

## Scheduler

### Cron

**Type:** `String`  
**Command Line:**  `-C, --cron`  
**Environment Variable:** `CRON`  
**Default**: `None`  
**Example:** `-e CRON="*/5 * * * *"`   

The schedule defined when to check for updates. If not defined, runs at interval.

### Interval

**Type:** `Integer`  
**Command Line:**  `-i, --interval`  
**Environment Variable:** `INTERVAL`
**Default**: `300`  
**Example:** `-e INTERVAL=300`  

The interval in seconds between checking for updates. There is a hard-coded 30 second minimum. Anything lower than that will set to 30. 

### Run Once

**Type:** `Boolean - Interrupting`  
**Command Line:**  `-o, --run-once`  
**Environment Variable:** `RUN_ONCE`  
**Default:** `False`  

Dockupdater will only do a single pass of all container checks, and then exit. This is a great way to granularly control scheduling with an outside scheduler like cron. If during the single pass dockupdater has to self-update, it will do another full pass after updating itself to ensure that all containers were checked.

## Docker Specifics

### Cleanup

**Type:** `Boolean`  
**Command Line:**  `-c, --cleanup`  
**Environment Variable:** `CLEANUP`  
**Default:** `False`  
**Availability:** `containers`  
**Override label:** [`dockupdater.cleamup`](Labels.md#cleanup)
**Example:** `-e CLEANUP=true`  

Remove the old images after updating. If you have multiple containers using the same image, it will ensure all containers are updated before successfully removing the image.

### Disable services check (swarm)

**Type:** `Boolean`  
**Command Line:**  `--disable-services-check`  
**Environment Variable:** `DISABLE_SERVICES_CHECK`  
**Default:** `False`  
**Availability:** `services`  
**Example:** `-e DISABLE_SERVICES_CHECK=true`  

Disable the scan for services (swarm). With this flag only standalone container will be updated.

### Disable containers check

**Type:** `Boolean`  
**Command Line:**  `--disable-containers-check`  
**Environment Variable:** `DISABLE_CONTAINERS_CHECK`  
**Default:** `False`  
**Availability:** `containers`  
**Example:** `-e DISABLE_CONTAINERS_CHECK=true`  

Disable the scan for standalone containers.

### Docker Socket

**Type:** `List - Multiple`
**Command Line:**  `-d, --docker-sockets`  
**Environment Variable:** `DOCKER_SOCKETS`  
**Default:** `unix://var/run/docker.sock`  
**Availability:** `containers` `services`  
**Example:** `-e DOCKER_SOCKETS="unix://var/run/docker.sock tcp://192.168.1.100:2376"`  

Allows you to define the docker socket.

### Docker TLS

**Type:** `Boolean`  
**Command Line:**  `-t, --docker-tls`  
**Environment Variable:** `DOCKER_TLS`  
**Default:** `False`  
**Availability:** `containers` `services`  
**Example:** `-e DOCKER_TLS=true -v $DOCKER_CERT_FOLDER:/root/.docker/`  

Enables docker TLS secure client connections by certificate

### Docker TLS Verify

**Type:** `Boolean`  
**Command Line:**  `-T, --docker-tls-verify`  
**Environment Variable:** `DOCKER_TLS_VERIFY`  
**Default:** `True`  
**Availability:** `containers` `services`  
**Example:** `-e DOCKER_TLS_VERIFY=false`  

Verify CA certificate for docker deamon

### Label

**Type:** `Boolean`  
**Command Line:**  `-k, --label`  
**Environment Variable:** `LABEL`  
**Default:** `False`  
**Availability:** `containers` `services`  
**Example:** `-e LABEL=true`  

This flag allows a more strict control over Dockupdater's updates. If the container or service does not have a `dockupdater.enable` label, it will be ignored completely.

### Latest

**Type:** `Boolean`  
**Command Line:**  `-L, --latest`  
**Environment Variable:** `LATEST`  
**Default:** `False`  
**Availability:** `containers` `services`  
**Override label:** [`dockupdater.latest`](Labels.md#latest)
**Example:** `-e LATEST=true`  

Pull the `:latest` tags and update all containers to it, regardless of the current tag the container is running as.

### Recreate first

**Type:** `Integer`  
**Command Line:**  `-F, --recreate-first`  
**Environment Variable:** `RECREATE_FIRST`  
**Default:** `0`  
**Availability:** `containers`  
**Override label:** [`dockupdater.recreate_first`](Labels.md#recreate-first)
**Example:** `-e RECREATE_FIRST=true`  

Work only with standalone container. To minimize application down time, we could create the new container before deleting the old.

Warning: This feature doesn't work if you have exposed ports. We highly recommend to use a load balancer like [Traefik](https://traefik.io/) if you need to use exposed ports with dockupdater.

Self update of dockupdater always use this feature.

### Repository User

**Type:** `String`  
**Command Line:**  `-r, --repo-user`  
**Environment Variable:** `REPO_USER`  
**Default:** `None`  
**Availability:** `containers` `services`  
**Example:** `-e REPO_USER=johndoe1970`  

Define a username for repository authentication. Will be ignored without defining a repository password.

### Repository Password

**Type:** `String`  
**Command Line:**  `-R, --repo-pass`  
**Environment Variable:** `REPO_PASS`  
**Default:** `None`  
**Availability:** `containers` `services`  
**Example:** `-e REPO_PASS=0791eodnhoj`  

Define a password for repository authentication. Will be ignored without defining a repository username.

### Stop signal

**Type:** `Integer`  
**Command Line:**  `--stop-signal`  
**Environment Variable:** `STOP_SIGNAL`  
**Default:** `None`  
**Availability:** `containers`  
**Override label:** [`dockupdater.stop_signal`](Labels.md#stop_signal)
**Example:** `-e STOP_SIGNAL=12`  

Define a stop signal to send to the container instead of SIGKILL. Default behavior is to use default docker stop command.

### Start

**Type:** `List String - Multiple/Space separated`  
**Command Line:**  `-S, --start`  
**Environment Variable:** `STARTS`  
**Default:** `None`  
**Availability:** `containers` `services`  
**Override label:** [`dockupdater.starts`](Labels.md#starts)
**Example:** `-e STARTS="weight:1,MyContainerName MyRegex[0-9]+"`  

> Warning: A bad configuration of this option may cause outage.

Define witch containers/services to start after an update. Docupdater will check for every container/service matching this name or [regex pattern](https://pythex.org/).

You can define a weight for each pattern. The container will start in the weight order, lower will start before higher. If the weight is omit, the value 100 is used.

When it use with **service**:

* If the service is stop by the option [`--stop`](#stop), the service will scale to the old number of replicas.
* If the service replicas is 0, Dockupdater will scale it to 1.
* If the service replicas is higher than 1, the service will be force updated. All containers in the service will restart.

A special token `{stack}` with the current stack name is available to controle witch service to start or stop. See the [Frequently asked questions](#Frequently-Asked-Questions.md#how-to-use-special-token-stack) for a complete example.

### Stop

**Type:** `List string - Multiple/Space separated`  
**Command Line:**  `-s, --stop`  
**Environment Variable:** `STOPS`  
**Default:** `0`  
**Availability:** `containers` `services`  
**Override label:** [`dockupdater.stops`](Labels.md#stops)
**Example:** `-e STOPS="weight:999,MyContainerName MyRegex[0-9]+"`  

Define witch containers/services to stop before an update. Docupdater will check for every container/service matching this name or [regex pattern](https://pythex.org/).

You can define a weight for each pattern. The container will stop in the weight order, lower will stop before higher. If the weight is omit, the value 100 is used.

On stop, service will scale to 0. If you use with option `--start`, on the start the replicas number will be restored. This doesn't work with service replicated in global mode.

See [`--start` option](#start) documentation for more example.

### Wait time

**Type:** `Integer`  
**Command Line:**  `-w, --wait`  
**Environment Variable:** `WAIT`  
**Default:** `0`  
**Availability:** `containers` `services`  
**Override label:** [`dockupdater.wait`](Labels.md#wait-time)
**Example:** `-e WAIT=60`  

Define a time in seconds to wait after an update before updating any others containers or services.

## Notifications

### Notifiers

**Type:** `List - Multiple/Space separated`  
**Command Line:**  `-N, --notifiers`  
**Environment Variable:** `NOTIFIERS`  
**Default:** `None`  
**Override label:** [`dockupdater.notifiers`](Labels.md#notifiers)
**Example:** `-e NOTIFIERS="mailtos://myUsername:myPassword@gmail.com?to=receivingAddress@gmail.com jsons://webhook.site/something"`  

Dockupdater uses [apprise](https://github.com/caronc/apprise) to support a large variety of notification platforms.

Notifications are sent for every update. The notification contains the container/service updated.

More information can be found in the [notifications docs](Notifications.md).

### Skip start notification

**Type:**  `Boolean`
**Command Line:**  `--skip-start-notif`  
**Environment Variable:** `SKIP_START_NOTIF`  
**Default:** `False`  
**Example:** `-e SKIP_START_NOTIF=true`  

Dockupdater send a notification when it start. That option disable this notification.

***

Next: [Customize usage with labels](Labels.md)

### Template file

**Type:** `Path`  
**Command Line:**  `--template-file`  
**Environment Variable:** `TEMPLATE_FILE`  
**Default:** `None`  
**Override label:** [`dockupdater.template_file`](Labels.md#template-file)
**Example:** `-e TEMPLATE_FILE="/template.j2"`  

Use this template instead of default template file. We use jinja2 template style to parse the template. Must be a valid path, don't forget to mount the template in the container.

See this example of template file:
```
{{ object.name }} ({{ object.get_image_name() }}:{{ object.get_tag() }}) updated from {{ object.get_current_id() }} to {{ object.get_latest_id() }}
```


docs/Timezone-Configuration.md
To more closely monitor ouroboros' actions and for accurate log ingestion, you can change the timezone of the container from UTC by setting the [`TZ`](http://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html) environment variable like so:

```
docker run -d --name ouroboros \
  -e TZ=America/Chicago \
  -v /var/run/docker.sock:/var/run/docker.sock \
  pyouroboros/ouroboros
```

CHANGELOG.md
# Change Log

## [1.2.1](https://github.com/pypyupdater/pyupdater/tree/1.2.1) (2019-02-13)
[Full Changelog](https://github.com/pypyupdater/pyupdater/compare/1.2.0...1.2.1)

**Fixed bugs:**

- Broken when no :tag specified  [\#210](https://github.com/pypyupdater/pyupdater/issues/210)

**Other Pull Requests**

- v1.2.1 Merge [\#213](https://github.com/pypyupdater/pyupdater/pull/213) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- v1.2.1 to develop [\#212](https://github.com/pypyupdater/pyupdater/pull/212) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- fixes \#210 [\#211](https://github.com/pypyupdater/pyupdater/pull/211) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- version bump to 1.2.1 + develop branch + twine fix + … [\#209](https://github.com/pypyupdater/pyupdater/pull/209) ([DirtyCajunRice](https://github.com/DirtyCajunRice))

## [1.2.0](https://github.com/pypyupdater/pyupdater/tree/1.2.0) (2019-02-14)
[Full Changelog](https://github.com/pypyupdater/pyupdater/compare/1.1.2...1.2.0)

**Implemented enhancements:**

- Move "Interval container update" messages to debug log level [\#194](https://github.com/pypyupdater/pyupdater/issues/194)
- \[Feature Request\] Support for Swarm Services [\#178](https://github.com/pypyupdater/pyupdater/issues/178)
- Add Warning for label\_enable not set while using labels\_only [\#202](https://github.com/pypyupdater/pyupdater/pull/202) ([larsderidder](https://github.com/larsderidder))

**Fixed bugs:**

- Change depends\_on logic [\#198](https://github.com/pypyupdater/pyupdater/issues/198)
- Containers relying upon network namespace of a container that gets updated breaks when the parent container is recreated [\#197](https://github.com/pypyupdater/pyupdater/issues/197)
- Exception when trying to update container with complex compose networks [\#196](https://github.com/pypyupdater/pyupdater/issues/196)
- Problem with network IPv4 address carry-over [\#193](https://github.com/pypyupdater/pyupdater/issues/193)
- Monitor Ignored Re-Address + jenkins cleanup [\#191](https://github.com/pypyupdater/pyupdater/pull/191) [[cleanup](https://github.com/pypyupdater/pyupdater/labels/cleanup)] ([DirtyCajunRice](https://github.com/DirtyCajunRice))

**Closed issues:**

- Remove legacy --latest [\#206](https://github.com/pypyupdater/pyupdater/issues/206) [[breaking change](https://github.com/pypyupdater/pyupdater/labels/breaking%20change)] [[cleanup](https://github.com/pypyupdater/pyupdater/labels/cleanup)]
- Add environment variables in Wiki [\#203](https://github.com/pypyupdater/pyupdater/issues/203) [[documentation](https://github.com/pypyupdater/pyupdater/labels/documentation)]
- Slack notifications via webhook not working [\#187](https://github.com/pypyupdater/pyupdater/issues/187)

**Other Pull Requests**

- v1.2.0 Merge [\#208](https://github.com/pypyupdater/pyupdater/pull/208) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- v1.2.0 to develop [\#207](https://github.com/pypyupdater/pyupdater/pull/207) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- Patch/tag bug [\#205](https://github.com/pypyupdater/pyupdater/pull/205) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- Patch/group 5 [\#201](https://github.com/pypyupdater/pyupdater/pull/201) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- Fix bug in user defined network detection [\#200](https://github.com/pypyupdater/pyupdater/pull/200) ([nightvisi0n](https://github.com/nightvisi0n))
- Adjust apscheduler logger [\#199](https://github.com/pypyupdater/pyupdater/pull/199) ([circa10a](https://github.com/circa10a))
- Carry over network config [\#195](https://github.com/pypyupdater/pyupdater/pull/195) ([nightvisi0n](https://github.com/nightvisi0n))
- Jenkins tweaks [\#192](https://github.com/pypyupdater/pyupdater/pull/192) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- Swarm + Jenkins [\#188](https://github.com/pypyupdater/pyupdater/pull/188) ([DirtyCajunRice](https://github.com/DirtyCajunRice))

## [1.1.2](https://github.com/pypyupdater/pyupdater/tree/1.1.2) (2019-02-02)
[Full Changelog](https://github.com/pypyupdater/pyupdater/compare/1.1.1...1.1.2)

**Fixed bugs:**

- No default timezone [\#176](https://github.com/pypyupdater/pyupdater/issues/176)

**Closed issues:**

- cron documentation example update [\#182](https://github.com/pypyupdater/pyupdater/issues/182) [[documentation](https://github.com/pypyupdater/pyupdater/labels/documentation)]

**Other Pull Requests**

- v1.1.2 Merge [\#186](https://github.com/pypyupdater/pyupdater/pull/186) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- v1.1.2 to develop [\#183](https://github.com/pypyupdater/pyupdater/pull/183) ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- Fix default timezone [\#177](https://github.com/pypyupdater/pyupdater/pull/177) ([circa10a](https://github.com/circa10a))

## [1.1.1](https://github.com/pypyupdater/pyupdater/tree/1.1.1) (2019-02-01)
[Full Changelog](https://github.com/pypyupdater/pyupdater/compare/1.1.0...1.1.1)

**Implemented enhancements:**

- Support for adding an identifier \(hostname?\) to notifications [\#158](https://github.com/pypyupdater/pyupdater/issues/158)
- Influx config data + ocd cleanup [\#162](https://github.com/pypyupdater/pyupdater/pull/162) [[cleanup](https://github.com/pypyupdater/pyupdater/labels/cleanup)] ([DirtyCajunRice](https://github.com/DirtyCajunRice))
- add cli arg for cron [\#157](https://github.com/pypyupdater/pyupdater/pull/157) ([DirtyCajunRice](https://github.com/DirtyCajunRice))

**Fixed bugs:**

- pyupdater does not respect MONITOR= [\#166](https://github.com/pypyupdater/pyupdater/issues/166)
- Docker TLS over TCP connections [\#154](https://github.com/pypyupdater/pyupdater/issues/154) [[documentation](https://github.com/pypyupdater/pyupdater/labels/documentation)]

dockupdater/lib/config.py
import re
from copy import deepcopy
from logging import getLogger
from os import environ
from pathlib import Path

from ..helpers.helpers import convert_to_boolean
from .logger import BlacklistFilter

OPTION_REGEX_PATTERN = r"^(?:weight:(?P<weight>\d+),)?(?P<regex>.*)$"
DEFAULT_REGEX_WEIGHT = 100

MINIMUM_INTERVAL = 30

ENABLE_LABEL = "dockupdater.enable"
DISABLE_LABEL = "dockupdater.disable"

LABELS_MAPPING = {
    "dockupdater.latest": "latest",
    "dockupdater.notifiers": "notifiers",
    "dockupdater.stop_signal": "stop_signal",
    "dockupdater.cleanup": "cleanup",
    "dockupdater.template_file": "template_file",
    "dockupdater.wait": "wait",
    "dockupdater.recreate_first": "recreate_first",
    "dockupdater.starts": "starts",
    "dockupdater.stops": "stops",
}


class DefaultConfig(object):
    """Default configuration"""
    hostname = environ.get('HOSTNAME')
    interval = 300
    cron = None
    docker_sockets = ['unix://var/run/docker.sock']
    docker_tls = False
    docker_tls_verify = True
    log_level = 'info'
    cleanup = False
    run_once = False
    label = False
    stop_signal = None
    disable_containers_check = False
    disable_services_check = False
    latest = False
    wait = 0
    recreate_first = False

    stops = []
    starts = []

    repo_user = None
    repo_pass = None

    notifiers = []
    skip_start_notif = False
    template_file = None


class OptionRegex(object):
    def __init__(self, pattern):
        """pattern may be in these format:
            - weight:Digit,regex_pattern
            - regex_pattern
        """
        match = re.fullmatch(OPTION_REGEX_PATTERN, pattern, re.IGNORECASE).groupdict()
        try:
            re.compile(match.get("regex"))  # Test the regex
            self.regex = match.get("regex")
        except re.error:
            raise AttributeError("Invalid regex {} for option start or stop.".format(match.get("regex")))
        self.weight = int(match.get("weight") or DEFAULT_REGEX_WEIGHT)
        self.tokens = None

    def match(self, name):
        regex = self.regex
        if self.tokens:
            for token, value in self.tokens.items():
                if token and value:
                    regex = regex.replace("{" + str(token) + "}", value)
        return bool(re.fullmatch(regex, name))

    def __repr__(self):
        return f"<Option {self.regex}[{self.weight}]>"


class Config(object):
    def __init__(self, **kwargs):
        super().__setattr__("options", kwargs)
        self.logger = getLogger()
        self.compute_args()
        self.filtered_strings = None

    @classmethod
    def from_labels(cls, config, labels):
        """Create a new config object from an existing config and a dict of docker labels"""
        options = deepcopy(config.options)

        if labels:
            for label, value in labels.items():
                if label in LABELS_MAPPING:
                    if label in ["dockupdater.notifiers"]:
                        options[LABELS_MAPPING[label]] = value.split(" ")
                    elif label in ["dockupdater.latest", "dockupdater.cleanup"]:
                        options[LABELS_MAPPING[label]] = convert_to_boolean(value)
                    elif label in ["dockupdater.wait"]:
                        options[LABELS_MAPPING[label]] = int(value)
                    elif label in ["dockupdater.stops", "dockupdater.starts"]:
                        options[LABELS_MAPPING[label]] = [OptionRegex(item) for item in value.split(" ")]
                    else:
                        options[LABELS_MAPPING[label]] = value
                    if label == "dockupdater.template_file":
                        # Reload template
                        options["template"] = Config.load_template(options.get('template_file'))
                elif label.startswith("dockupdater."):
                    config.logger.warning("Warning label %s doesn't exist", label)

        return cls(**options)

    def __setattr__(self, key, value):
        if key in self.options:
            self.options[key] = value
        else:
            super().__setattr__(key, value)

    def __getattr__(self, attr):
        try:
            return self.options[attr]
        except KeyError:
            raise AttributeError

    def config_blacklist(self):
        """Mask sensitive data from logs"""
        filtered_strings = [
            getattr(self, key.lower()) for key, value in self.options.items()
            if key.lower() in BlacklistFilter.blacklisted_keys
        ]
        # Clear None values
        self.filtered_strings = list(filter(None, filtered_strings))
        # take lists inside of list and append to list
        for index, value in enumerate(self.filtered_strings, 0):
            if isinstance(value, list) or isinstance(value, tuple):
                self.filtered_strings.extend(self.filtered_strings.pop(index))
                self.filtered_strings.insert(index, self.filtered_strings[-1:][0])
        # Filter out no string item
        self.filtered_strings = [item for item in self.filtered_strings if isinstance(item, str)]
        # Added matching for ports
        ports = [string.split(':')[0] for string in self.filtered_strings if ':' in string]
        self.filtered_strings.extend(ports)
        # Added matching for tcp sockets. ConnectionPool ignores the tcp://
        tcp_sockets = [string.split('//')[1] for string in self.filtered_strings if '//' in string]
        self.filtered_strings.extend(tcp_sockets)
        # Get JUST hostname from tcp//unix
        for socket in getattr(self, 'docker_sockets'):
            self.filtered_strings.append(socket.split('//')[1].split(':')[0])

        for handler in self.logger.handlers:
            handler.addFilter(BlacklistFilter(set(self.filtered_strings)))

    def compute_args(self):
        if self.repo_user and self.repo_pass:
            self.options['auth_json'] = {'Username': self.repo_user, 'Password': self.repo_pass}
        else:
            self.options['auth_json'] = None

        if self.disable_containers_check and self.disable_services_check:
            raise AttributeError("Error you can't disable all monitoring (containers/services).")

        # Config sanity checks
        if self.cron:
            if not isinstance(self.cron, list):
                cron_times = self.cron.strip().split(' ')
                if len(cron_times) != 5:
                    self.logger.critical("Cron must be in cron syntax. e.g. * * * * * (5 places).")
                    raise AttributeError("Invalid cron")
                else:
                    self.logger.info("Cron configuration is valid. Using Cron schedule %s", cron_times)
                    self.cron = cron_times
                    self.interval = None
        else:
            if self.interval < MINIMUM_INTERVAL:
                self.logger.warning('Minimum value for interval was 30 seconds.')
                self.interval = MINIMUM_INTERVAL

        # Convert parameters to regex object
        self.stops = [OptionRegex(stop) if not isinstance(stop, OptionRegex) else stop for stop in self.stops]
        self.stops.sort(key=lambda x: x.weight)
        self.starts = [OptionRegex(start) if not isinstance(start, OptionRegex) else start for start in self.starts]
        self.starts.sort(key=lambda x: x.weight)

        self.options['template'] = Config.load_template(self.template_file)

    @staticmethod
    def load_template(template_file):
        # Load default template file
        if not template_file:
            dir_path = Path().absolute()
            template_file = dir_path.joinpath("dockupdater/templates/notification.j2")

        if Path(template_file).exists():
            with open(template_file) as f:
                return f.read()
        else:
            raise AttributeError(f"Template file {template_file} not found")
